module-system/module-implementations/sov-blob-storage/src/capabilities.rs

```diff
@@ -0,0 +1,88 @@
+use sov_modules_api::capabilities::{BlobRefOrOwned, BlobSelector};
+use sov_modules_api::{Context, Spec};
+use sov_rollup_interface::da::BlobReaderTrait;
+use sov_state::WorkingSet;
+use tracing::info;
+
+use crate::BlobStorage;
+
+impl<C: Context> BlobSelector for BlobStorage<C> {
+    type Context = C;
+
+    fn get_blobs_for_this_slot<'a, I, B>(
+        &self,
+        current_blobs: I,
+        working_set: &mut WorkingSet<<Self::Context as Spec>::Storage>,
+    ) -> anyhow::Result<Vec<BlobRefOrOwned<'a, B>>>
+    where
+        B: BlobReaderTrait,
+        I: IntoIterator<Item = &'a mut B>,
+    {
+        // TODO: Chain-state module: https://github.com/Sovereign-Labs/sovereign-sdk/pull/598/
+        let current_slot: u64 = self.get_current_slot_number(working_set);
+        let past_deferred: Vec<B> = current_slot
+            .checked_sub(self.get_deferred_slots_count(working_set))
+            .map(|pull_from_slot| self.take_blobs_for_block_number(pull_from_slot, working_set))
+            .unwrap_or_default();
+        let preferred_sequencer = self.get_preferred_sequencer(working_set);
+
+        let preferred_sequencer = if let Some(sequencer) = preferred_sequencer {
+            sequencer
+        } else {
+            return Ok(past_deferred
+                .into_iter()
+                .map(Into::into)
+                .chain(current_blobs.into_iter().map(Into::into))
+                .collect());
+        };
+
+        let mut priority_blobs = Vec::new();
+        let mut to_defer: Vec<&mut B> = Vec::new();
+
+        for blob in current_blobs {
+            if blob.sender().as_ref() == &preferred_sequencer[..] {
+                priority_blobs.push(blob);
+            } else {
+                to_defer.push(blob);
+            }
+        }
+
+        // TODO: chain state module: https://github.com/Sovereign-Labs/sovereign-sdk/pull/598/
+        self.slot_number.set(&(current_slot + 1), working_set);
+
+        if !to_defer.is_empty() {
```

Comment:
nit: I think this if is unnecessary since we use iterator combinators everywhere

Reply:
but we allocate inside.

It is single reference, but still an allocation. Also if we have any other logic related to saving, it is inside this branch.

Previous messages made me think about potential attach vector.

For simplicity, imagine there are preferred sequencer, but he is inactive at the moment. So all blobs just moved to next blob.

Malicious sequencer can submit 3 blobs: A, B and C in single slot number 100. 

The sequencer is registered at the beginning of the slot 100. But blob A contains exit transaction, which makes blobs B and C to be sent from unregistered sequencer.

ApplySlotHooks only run once per slot and all blobs are checked against the state at the beginning of the slot. Slo ApplySlotHooks will save all 3 blobs: A, B and C to be executed in slot 101, but blobs B and C will be discarded by `ApplyBlobHooks`. Considering that storing blobs is expensive operation, this might be costly.